geo/src/algorithm/extremes.rs

```diff
@@ -1,91 +1,87 @@
-use super::kernels::HasKernel;
-use crate::algorithm::convex_hull::ConvexHull;
-use crate::{ExtremePoint, Extremes};
-use crate::{MultiPoint, MultiPolygon, Point, Polygon};
-use num_traits::{Float, Signed};
+use super::kernels::*;
+use crate::prelude::*;
+use crate::*;
+use num_traits::{Signed, Zero};
+
+/// Compute the sign of the dot product of `u` and `v` using
+/// robust predicates. The output is `CounterClockwise` if
+/// the sign is positive, `Clockwise` if negative, and
+/// `Collinear` if zero.
+fn dot_product_sign<T>(u: Coordinate<T>, v: Coordinate<T>) -> Orientation
+where
+    T: CoordinateType + Signed + HasKernel,
+{
+    let zero = Coordinate::zero();
+    let vdash = Coordinate { x: -v.y, y: v.x };
+    T::Ker::orient2d(zero, u, vdash)
+}
 
 // Useful direction vectors, aligned with x and y axes:
 // 1., 0. = largest x
 // 0., 1. = largest y
 // 0., -1. = smallest y
 // -1, 0. = smallest x
 
-// various tests for vector orientation relative to a direction vector u
-
-// Not currently used, but maybe useful in the future
-#[allow(dead_code)]
-fn up<T>(u: Point<T>, v: Point<T>) -> bool
+/// Predicate that returns `true` if `vi` is (strictly) above `vj`
+/// along the direction of `u`
+fn above<T>(u: Coordinate<T>, vi: Coordinate<T>, vj: Coordinate<T>) -> bool
 where
-    T: Float,
+    T: CoordinateType + Signed + HasKernel,
 {
-    u.dot(v) > T::zero()
+    dot_product_sign(u, vi - vj) == Orientation::CounterClockwise
 }
 
-fn direction_sign<T>(u: Point<T>, vi: Point<T>, vj: Point<T>) -> T
-where
-    T: Float,
-{
-    u.dot(vi - vj)
-}
-
-// true if Vi is above Vj
-fn above<T>(u: Point<T>, vi: Point<T>, vj: Point<T>) -> bool
-where
-    T: Float,
-{
-    direction_sign(u, vi, vj) > T::zero()
-}
-
-// true if Vi is below Vj
-// Not currently used, but maybe useful in the future
+/// Predicate that returns `true` if `vi` is (strictly) below `vj`
+/// along the direction of `u`
 #[allow(dead_code)]
-fn below<T>(u: Point<T>, vi: Point<T>, vj: Point<T>) -> bool
+fn below<T>(u: Coordinate<T>, vi: Coordinate<T>, vj: Coordinate<T>) -> bool
 where
-    T: Float,
+    T: CoordinateType + Signed + HasKernel,
 {
-    direction_sign(u, vi, vj) < T::zero()
+    dot_product_sign(u, vi - vj) == Orientation::Clockwise
 }
 
 // wrapper for extreme-finding function
 fn find_extreme_indices<T, F>(func: F, polygon: &Polygon<T>) -> Result<Extremes, ()>
 where
-    T: Float + Signed,
-    F: Fn(Point<T>, &Polygon<T>) -> Result<usize, ()>,
+    T: CoordinateType + HasKernel + Signed,
+    F: Fn(Coordinate<T>, &Polygon<T>) -> Result<usize, ()>,
 {
-    if !polygon.is_convex() {
+    use crate::is_convex::IsConvex;
+    if !polygon.exterior().is_convex() {
         return Err(());
     }
-    let directions = vec![
-        Point::new(T::zero(), -T::one()),
-        Point::new(T::one(), T::zero()),
-        Point::new(T::zero(), T::one()),
-        Point::new(-T::one(), T::zero()),
+    let directions: Vec<Coordinate<_>> = vec![
```

Comment:
no biggie so take it or leave it, but since this is fixed size we could make this an array instead of a vec.

Reply:
I don't think fixed size works here because it is not an iterator.  We can re-visit this once we have the `ForEachCoord` trait, and simplify the logic here.